### PR TITLE
Choice operator on Java8 branch

### PR DESCRIPTION
**Pull request by agwe**

### CODE REVIEW

File: jolie/src/jolie/OOITBuilder.java [hunk 1]
```diff
@@ -1384,22 +1239,16 @@ public void visit( InstanceOfExpressionNode n )
 	public void visit( TypeCastExpressionNode n )
```
Comment (fmontesi) [inline jolie/src/jolie/OOITBuilder.java]: Looks like the commit is reverting my changes to this method for some reason. The correct version is the one with the switch.

File: jolie/src/jolie/OOITBuilder.java [hunk 2]
```diff
@@ -1670,6 +1519,47 @@ public void visit( ProvideUntilStatement n )
...
 	}
 
+	@Override
+	public void visit(TypeChoiceDefinition n) {
+		Type currType1;
+		Type currType2;
+		boolean backupInsideType = false;
+		insideType = true;
+
+		if ( n.untypedSubTypes() ) {
```
Comment (fmontesi) [inline jolie/src/jolie/OOITBuilder.java]: For building currType1 and currType2, can't you just recursively call this visitor (OOITBuilder) ?

File: jolie/src/jolie/OOITBuilder.java [hunk 3]
```diff
@@ -1670,6 +1519,47 @@ public void visit( ProvideUntilStatement n )
...
+				for( Entry< String, TypeDefinition > entry : n.t2SubTypes() ) {
+					subTypes.put( entry.getKey(), buildType( entry.getValue() ) );
+				}
+			}
+			currType2 = Type.create( n.t2NativeType(), n.getT2Cardinality(), false, subTypes );
+		}
+
+		insideType = backupInsideType;
+
+		if ( insideType == false && insideOperationDeclaration == false ) {
```
Comment (fmontesi) [inline jolie/src/jolie/OOITBuilder.java]: What about the other possibility? (insideType == true)

File: jolie/src/jolie/runtime/typing/Type.java [hunk 4]
```diff
@@ -1,5 +1,5 @@
 /***************************************************************************
- *   Copyright (C) 2009-2015 by Fabrizio Montesi <famontesi@gmail.com>     *
+ *   Copyright (C) 2009-2011 by Fabrizio Montesi <famontesi@gmail.com>     *
```
Comment (fmontesi) [inline jolie/src/jolie/runtime/typing/Type.java]: Should not be reverted.

File: jolie/src/jolie/runtime/typing/Type.java [hunk 5]
```diff
@@ -50,20 +51,18 @@ public TypeImpl(
 			this.subTypes = Collections.unmodifiableMap( subTypes );
 		}
 	}
-	
-	@Override
```
Comment (fmontesi) [inline jolie/src/jolie/runtime/typing/Type.java]: Should not be reverted.

File: jolie/src/jolie/runtime/typing/Type.java [hunk 6]
```diff
@@ -50,20 +51,18 @@ public TypeImpl(
...
+
+
 	public Map< String, Type > subTypes()
 	{
 		return subTypes;
 	}
 
-	@Override
```
Comment (fmontesi) [inline jolie/src/jolie/runtime/typing/Type.java]: Should not be reverted.

File: jolie/src/jolie/runtime/typing/Type.java [hunk 7]
```diff
@@ -50,20 +51,18 @@ public TypeImpl(
...
 	public Range cardinality()
 	{
 		return cardinality;
 	}
 	
-	@Override
```
Comment (fmontesi) [inline jolie/src/jolie/runtime/typing/Type.java]: Should not be reverted.

File: jolie/src/jolie/runtime/typing/Type.java [hunk 8]
```diff
@@ -73,7 +72,6 @@ public void cutChildrenFromValue( Value value )
 		}
 	}
 
-	@Override
```
Comment (fmontesi) [inline jolie/src/jolie/runtime/typing/Type.java]: Should not be reverted.

File: jolie/src/jolie/runtime/typing/Type.java [hunk 9]
```diff
@@ -93,13 +91,13 @@ private void castSubType( String typeName, Type type, Value value, StringBuilder
 		pathBuilder.append( '.' );
 		pathBuilder.append( typeName );
 
-		final boolean hasChildren = value.hasChildren( typeName );
-		if ( hasChildren == false && type.cardinality().min() > 0 ) {
+		boolean hasChildren = value.hasChildren( typeName );
```
Comment (fmontesi) [inline jolie/src/jolie/runtime/typing/Type.java]: Should be final.

File: jolie/src/jolie/runtime/typing/Type.java [hunk 10]
```diff
@@ -93,13 +91,13 @@ private void castSubType( String typeName, Type type, Value value, StringBuilder
...
+		if ( hasChildren == false && type.cardinality().min() > 0 && !(type instanceof TypePair) ) {
 			throw new TypeCastingException( "Undefined required child node: " + pathBuilder.toString() );
 		} else if ( hasChildren ) {
-			final ValueVector vector = value.getChildren( typeName );
-			final int size = vector.size();
-			if ( type.cardinality().min() > size || type.cardinality().max() < size ) {
+			ValueVector vector = value.getChildren( typeName );
```
Comment (fmontesi) [inline jolie/src/jolie/runtime/typing/Type.java]: Should be final.

File: jolie/src/jolie/runtime/typing/Type.java [hunk 11]
```diff
@@ -93,13 +91,13 @@ private void castSubType( String typeName, Type type, Value value, StringBuilder
...
+			int size = vector.size();
```
Comment (fmontesi) [inline jolie/src/jolie/runtime/typing/Type.java]: Should be final.

File: jolie/src/jolie/runtime/typing/Type.java [hunk 12]
```diff
@@ -112,7 +110,6 @@ private void castSubType( String typeName, Type type, Value value, StringBuilder
 		}
 	}
 
-	@Override
```
Comment (fmontesi) [inline jolie/src/jolie/runtime/typing/Type.java]: Should not be reverted.

File: jolie/src/jolie/runtime/typing/Type.java [hunk 13]
```diff
@@ -121,12 +118,9 @@ protected void check( Value value, StringBuilder pathBuilder )
 		}
 
 		if ( subTypes != null ) {
-			final int l = pathBuilder.length();
 			for( Entry< String, Type > entry : subTypes.entrySet() ) {
-				checkSubType( entry.getKey(), entry.getValue(), value, pathBuilder );
-				pathBuilder.setLength( l );
+				checkSubType( entry.getKey(), entry.getValue(), value, new StringBuilder( pathBuilder ) );
```
Comment (fmontesi) [inline jolie/src/jolie/runtime/typing/Type.java]: Lines 124, 126, and 127 should not be reverted.

File: jolie/src/jolie/runtime/typing/Type.java [hunk 14]
```diff
@@ -136,19 +130,21 @@ protected void check( Value value, StringBuilder pathBuilder )
 		}
 	}
 
-	private void checkSubType( String typeName, Type type, Value value, StringBuilder pathBuilder )
+	private void checkSubType(String typeName, Type type, Value value, StringBuilder pathBuilder)
```
Comment (fmontesi) [inline jolie/src/jolie/runtime/typing/Type.java]: Previous whitespace was correct.

File: jolie/src/jolie/runtime/typing/Type.java [hunk 15]
```diff
@@ -73,7 +72,6 @@ public void cutChildrenFromValue( Value value )
 		}
 	}
 
-	@Override
 	protected Value cast( Value value, StringBuilder pathBuilder )
```
Comment (fmontesi) [inline jolie/src/jolie/runtime/typing/Type.java]: The best way to implement this is probably to re-implement cast from TypePair, which could then call cast at the first type and then, if that fails, cast at the second type? The same goes for method check.

File: jolie/src/jolie/runtime/typing/Type.java [hunk 16]
```diff
@@ -222,29 +222,27 @@ private void castNativeType( Value value, StringBuilder pathBuilder )
 
 	private boolean checkNativeType( Value value, NativeType nativeType )
 	{
-		switch( nativeType ) {
```
Comment (fmontesi) [inline jolie/src/jolie/runtime/typing/Type.java]: This method should not be reverted (the switch version is correct).

File: jolie/src/jolie/runtime/typing/Type.java [hunk 17]
```diff
@@ -222,29 +222,27 @@ private void castNativeType( Value value, StringBuilder pathBuilder )
...
-		case RAW:
+		} else if ( nativeType == NativeType.RAW ) {
 			return value.isByteArray();
 		}
-		
+
 		return false;
 	}
 	
-	@Override
```
Comment (fmontesi) [inline jolie/src/jolie/runtime/typing/Type.java]: Should not be reverted.

File: jolie/src/jolie/runtime/typing/Type.java [hunk 18]
```diff
@@ -276,10 +274,10 @@ public static TypeLink createLink( String linkedTypeName, Range cardinality )
 	
 	public static Type merge( Type t1, Type t2 )
 	{
```
Comment (fmontesi) [inline jolie/src/jolie/runtime/typing/Type.java]: This method shuold not be reverted.

File: jolie/src/jolie/runtime/typing/Type.java [hunk 19]
```diff
@@ -323,13 +321,11 @@ public TypeLink( String linkedTypeName, Range cardinality )
 			this.cardinality = cardinality;
 		}
 		
-		@Override
```
Comment (fmontesi) [inline jolie/src/jolie/runtime/typing/Type.java]: Should not be reverted.

File: jolie/src/jolie/runtime/typing/Type.java [hunk 20]
```diff
@@ -323,13 +321,11 @@ public TypeLink( String linkedTypeName, Range cardinality )
...
 		public Map< String, Type > subTypes()
 		{
 			return linkedType.subTypes();
 		}
 		
-		@Override
```
Comment (fmontesi) [inline jolie/src/jolie/runtime/typing/Type.java]: Should not be reverted.

File: jolie/src/jolie/runtime/typing/Type.java [hunk 21]
```diff
@@ -344,27 +340,23 @@ public void setLinkedType( Type linkedType )
 		{
 			this.linkedType = linkedType;
 		}
-		
-		@Override
```
Comment (fmontesi) [inline jolie/src/jolie/runtime/typing/Type.java]: Should not be reverted.

File: jolie/src/jolie/runtime/typing/Type.java [hunk 22]
```diff
@@ -344,27 +340,23 @@ public void setLinkedType( Type linkedType )
...
+
 		public void cutChildrenFromValue( Value value )
 		{
 			linkedType.cutChildrenFromValue( value );
 		}
 
-		@Override
```
Comment (fmontesi) [inline jolie/src/jolie/runtime/typing/Type.java]: Should not be reverted.

File: jolie/src/jolie/runtime/typing/Type.java [hunk 23]
```diff
@@ -344,27 +340,23 @@ public void setLinkedType( Type linkedType )
...
 		public Range cardinality()
 		{
 			return cardinality;
 		}
 
-		@Override
```
Comment (fmontesi) [inline jolie/src/jolie/runtime/typing/Type.java]: Should not be reverted.

File: jolie/src/jolie/runtime/typing/Type.java [hunk 24]
```diff
@@ -344,27 +340,23 @@ public void setLinkedType( Type linkedType )
...
 		protected void check( Value value, StringBuilder pathBuilder )
 			throws TypeCheckingException
 		{
 			linkedType.check( value, pathBuilder );
 		}
 
-		@Override
```
Comment (fmontesi) [inline jolie/src/jolie/runtime/typing/Type.java]: Should not be reverted.

File: jolie/src/jolie/runtime/typing/TypePair.java [hunk 25]
```diff
@@ -0,0 +1,84 @@
...
+import jolie.lang.NativeType;
+import jolie.runtime.Value;
+import jolie.util.Range;
+
+import java.util.Map;
+
+/**
+ * Created by User on 01.08.2015.
+ */
+public class TypePair extends Type {
```
Comment (fmontesi) [inline jolie/src/jolie/runtime/typing/TypePair.java]: TypePair is a bit unintuitive. Either TypeChoice or TypeSum would be more standard.

File: jolie/src/jolie/runtime/typing/TypePair.java [hunk 26]
```diff
@@ -0,0 +1,84 @@
...
+    public Type getT1() {
```
Comment (fmontesi) [inline jolie/src/jolie/runtime/typing/TypePair.java]: How about left() and right() instead of getT1 and getT2? This way, we get an intuitive name that maps to the syntax (where we have a left and right type).

File: jolie/src/jolie/runtime/typing/TypePair.java [hunk 27]
```diff
@@ -0,0 +1,84 @@
...
+        this.t1 = t1;
+    }
+
+    @Override
+    public void cutChildrenFromValue(Value value) {
+
+    }
+
+    @Override
+    public NativeType nativeType() {
```
Comment (fmontesi) [inline jolie/src/jolie/runtime/typing/TypePair.java]: Ah, this explains why you have all those "instanceof" checks in the previous code.

I think this shows that we should not have nativeType(), cardinality(), and subTypes() as public methods in Type, since they do not make sense in TypePair. Otherwise we get a leaky abstraction.

File: jolie/src/jolie/runtime/typing/TypePair.java [hunk 28]
```diff
@@ -0,0 +1,84 @@
...
+    public Map<String, Type> subTypes() {
+        return null;
+    }
+
+    @Override
+    protected void check(Value value, StringBuilder pathBuilder) throws TypeCheckingException {
+        Boolean isValid;
+        isValid = isCheck(this.getT1(), value, pathBuilder) || isCheck(this.getT2(), value, pathBuilder);
+        if (!isValid){
+            throw new TypeCheckingException( "Invalid types for node " + pathBuilder.toString() + ": expected " + this.getT1().nativeType() + " or " +  this.getT2().nativeType() + ", found " + (( value.valueObject() == null ) ? "void" : value.valueObject().getClass().getName()) );
```
Comment (fmontesi) [inline jolie/src/jolie/runtime/typing/TypePair.java]: How can you be sure that the error came from a native type? Couldn't it be a nested sub node?

File: jolie/src/jolie/runtime/typing/TypePair.java [hunk 29]
```diff
@@ -0,0 +1,84 @@
...
+        }
+    }
+
+    @Override
+    protected Value cast(Value value, StringBuilder pathBuilder) throws TypeCastingException {
+        Boolean isValid;
+        isValid = isCast(this.getT1(), value, pathBuilder) || isCast(this.getT2(), value, pathBuilder);
+        if (!isValid){
+            throw new TypeCastingException( "Invalid types for node " + pathBuilder.toString() + ": expected " + this.getT1().nativeType() + " or " +  this.getT2().nativeType() + ", found " + (( value.valueObject() == null ) ? "void" : value.valueObject().getClass().getName()) );
```
Comment (fmontesi) [inline jolie/src/jolie/runtime/typing/TypePair.java]: Same comment as before, for check.

File: libjolie/src/jolie/lang/parse/OLVisitor.java [hunk 30]
```diff
@@ -21,159 +21,168 @@
...
-	public void visit( FreshValueExpressionNode n );
-	public void visit( CourierDefinitionNode n );
-	public void visit( CourierChoiceStatement n );
-	public void visit( NotificationForwardStatement n );
-	public void visit( SolicitResponseForwardStatement n );
-	public void visit( InterfaceExtenderDefinition n );
-	public void visit( InlineTreeExpressionNode n );
-	public void visit( VoidExpressionNode n );
-	public void visit( ProvideUntilStatement n );
+public interface OLVisitor {
```
Comment (fmontesi) [inline libjolie/src/jolie/lang/parse/OLVisitor.java]: Huh? The whole file changed but the change is actually minimal. Probably because of the different code styling? Can you use the code style we are currently using?

File: libjolie/src/jolie/lang/parse/SemanticVerifier.java [hunk 31]
```diff
@@ -430,7 +366,53 @@ public void visit( TypeDefinitionLink n )
 		definedTypeLinks.add( n );
 	}
 
-	private void checkCardinality( TypeDefinition type )
+	public void visit(TypeChoiceDefinition n) {
+		/*ArrayList<TypeDefinition> typ = new ArrayList<TypeDefinition>();
```
Comment (fmontesi) [inline libjolie/src/jolie/lang/parse/SemanticVerifier.java]: What's this commented code for?

File: libjolie/src/jolie/lang/parse/SemanticVerifier.java [hunk 32]
```diff
@@ -430,7 +366,53 @@ public void visit( TypeDefinitionLink n )
...
+			}
+		}
+
+		isTopLevelType = backupRootType;
+
+		if ( isTopLevelType ) {
+			definedTypes.put( n.id(), n );
+		}
+
+        if (n.getT1() instanceof TypeDefinitionLink){
```
Comment (fmontesi) [inline libjolie/src/jolie/lang/parse/SemanticVerifier.java]: Weird indentation here.

File: libjolie/src/jolie/lang/parse/SemanticVerifier.java [hunk 33]
```diff
@@ -277,11 +203,21 @@ private void resolveLazyLinks()
 		for( TypeDefinitionLink l : definedTypeLinks ) {
 			l.setLinkedType( definedTypes.get( l.linkedTypeName() ) );
 			if ( l.linkedType() == null ) {
-				error( l, "type " + l.id() + "points to an undefined type (" + l.linkedTypeName() + ")" );
+				error( l, "type " + l.id() + " points to an undefined type (" + l.linkedTypeName() + ")" );
 			}
 		}
 	}
 
+    private void resolveChoiceLazyLinks()
```
Comment (fmontesi) [inline libjolie/src/jolie/lang/parse/SemanticVerifier.java]: What is this method for? Doesn't the previous one already do this?

File: libjolie/src/jolie/lang/parse/ast/types/TypeChoiceDefinition.java [hunk 34]
```diff
@@ -0,0 +1,96 @@
...
+import java.util.Set;
+
+public class TypeChoiceDefinition extends TypeDefinition {
+    private final TypeDefinition T1;
+    private TypeDefinition T2;
+    //private Map< String, TypeDefinition > subTypes = null;
+    //private boolean untypedSubTypes = false;
+    private Range t2Cardinality;
+
+    public TypeChoiceDefinition(ParsingContext context, String id, Range cardinality, TypeDefinition T1)
```
Comment (fmontesi) [inline libjolie/src/jolie/lang/parse/ast/types/TypeChoiceDefinition.java]: Should take both t1 and t2 in the constructor.

File: libjolie/src/jolie/lang/parse/ast/types/TypeChoiceDefinition.java [hunk 35]
```diff
@@ -0,0 +1,96 @@
...
+import jolie.lang.NativeType;
+import jolie.lang.parse.OLVisitor;
+import jolie.lang.parse.context.ParsingContext;
+import jolie.util.Range;
+
+import java.util.Map;
+import java.util.Set;
+
+public class TypeChoiceDefinition extends TypeDefinition {
+    private final TypeDefinition T1;
```
Comment (fmontesi) [inline libjolie/src/jolie/lang/parse/ast/types/TypeChoiceDefinition.java]: Variables should not start with an upper case. I propose "left", and "right" instead of T2.

File: libjolie/src/jolie/lang/parse/ast/types/TypeChoiceDefinition.java [hunk 36]
```diff
@@ -0,0 +1,96 @@
...
+    private TypeDefinition T2;
```
Comment (fmontesi) [inline libjolie/src/jolie/lang/parse/ast/types/TypeChoiceDefinition.java]: Shuold be final once we get it from the constructor.

File: libjolie/src/jolie/lang/parse/ast/types/TypeChoiceDefinition.java [hunk 37]
```diff
@@ -0,0 +1,96 @@
...
+    //private boolean untypedSubTypes = false;
+    private Range t2Cardinality;
+
+    public TypeChoiceDefinition(ParsingContext context, String id, Range cardinality, TypeDefinition T1)
+    {
+        super(context, id, cardinality );
+        this.T1 = T1;
+    }
+
+    public TypeChoiceDefinition setT2(TypeDefinition T2, Range cardinality) {
```
Comment (fmontesi) [inline libjolie/src/jolie/lang/parse/ast/types/TypeChoiceDefinition.java]: Shuold be removed once we get T2 from the constructor.

File: libjolie/src/jolie/lang/parse/ast/types/TypeChoiceDefinition.java [hunk 38]
```diff
@@ -0,0 +1,96 @@
...
+        this.T2 = T2;
+        this.t2Cardinality = cardinality;
+        return this;
+    }
+
+    public Range getT2Cardinality(){
```
Comment (fmontesi) [inline libjolie/src/jolie/lang/parse/ast/types/TypeChoiceDefinition.java]: Why is this needed? Can't we just call getT2.cardinality() ?

The same goes for all other methods accessing T1 and T2.

File: libjolie/src/jolie/lang/parse/ast/types/TypeChoiceDefinition.java [hunk 39]
```diff
@@ -0,0 +1,96 @@
...
+    @Override
+    public boolean hasSubTypes() {
+        if (T1!=null) {
+            return T1.hasSubTypes();
+        }
+        else return false;
+    }
+
+    @Override
+    public boolean untypedSubTypes() {
```
Comment (fmontesi) [inline libjolie/src/jolie/lang/parse/ast/types/TypeChoiceDefinition.java]: I don't think we should expose this, it doesn't make much sense. Maybe we should not make untypedSubTypes() a public method in TypeDefinition anymore.

File: libjolie/src/jolie/lang/parse/ast/types/TypeDefinitionLink.java [hunk 40]
```diff
@@ -65,43 +66,39 @@ public TypeDefinition linkedType()
 		return linkedType;
 	}
 
-	@Override
```
Comment (fmontesi) [inline libjolie/src/jolie/lang/parse/ast/types/TypeDefinitionLink.java]: Should not be reverted.

File: libjolie/src/jolie/lang/parse/ast/types/TypeDefinitionLink.java [hunk 41]
```diff
@@ -65,43 +66,39 @@ public TypeDefinition linkedType()
...
 	{
 		return linkedType.untypedSubTypes();
 	}
 
-	@Override
-	public boolean hasSubTypes()
-	{
-		return linkedType.hasSubTypes();
-	}
+    public boolean hasSubTypes()
```
Comment (fmontesi) [inline libjolie/src/jolie/lang/parse/ast/types/TypeDefinitionLink.java]: Weird indentation here.

File: libjolie/src/jolie/lang/parse/ast/types/TypeDefinitionLink.java [hunk 42]
```diff
@@ -65,43 +66,39 @@ public TypeDefinition linkedType()
...
+    {
```
Comment (fmontesi) [inline libjolie/src/jolie/lang/parse/ast/types/TypeDefinitionLink.java]: How can the linkedType be null?

File: libjolie/src/jolie/lang/parse/ast/types/TypeDefinitionLink.java [hunk 43]
```diff
@@ -65,43 +66,39 @@ public TypeDefinition linkedType()
...
+        if (linkedType!=null) {
+            return linkedType.hasSubTypes();
+        }
+        else return false;
+    }
 
-	@Override
```
Comment (fmontesi) [inline libjolie/src/jolie/lang/parse/ast/types/TypeDefinitionLink.java]: Should not be reverted.

File: libjolie/src/jolie/lang/parse/ast/types/TypeDefinitionLink.java [hunk 44]
```diff
@@ -65,43 +66,39 @@ public TypeDefinition linkedType()
...
 	public TypeDefinition getSubType( String id )
 	{
 		return linkedType.getSubType( id );
 	}
 
-	@Override
```
Comment (fmontesi) [inline libjolie/src/jolie/lang/parse/ast/types/TypeDefinitionLink.java]: Should not be reverted.

File: libjolie/src/jolie/lang/parse/ast/types/TypeDefinitionLink.java [hunk 45]
```diff
@@ -65,43 +66,39 @@ public TypeDefinition linkedType()
...
 	public NativeType nativeType()
 	{
 		return linkedType.nativeType();
 	}
 
-	@Override
```
Comment (fmontesi) [inline libjolie/src/jolie/lang/parse/ast/types/TypeDefinitionLink.java]: Should not be reverted.

File: libjolie/src/jolie/lang/parse/ast/types/TypeDefinitionLink.java [hunk 46]
```diff
@@ -65,43 +66,39 @@ public TypeDefinition linkedType()
...
 	public Set< Map.Entry< String, TypeDefinition > > subTypes()
 	{
 		return linkedType.subTypes();
 	}
 
-	@Override
```
Comment (fmontesi) [inline libjolie/src/jolie/lang/parse/ast/types/TypeDefinitionLink.java]: Should not be reverted.

File: libjolie/src/jolie/lang/parse/ast/types/TypeDefinitionLink.java [hunk 47]
```diff
@@ -65,43 +66,39 @@ public TypeDefinition linkedType()
...
 	public boolean hasSubType( String id )
 	{
 		return linkedType.hasSubType( id );
 	}
 
-	@Override
```
Comment (fmontesi) [inline libjolie/src/jolie/lang/parse/ast/types/TypeDefinitionLink.java]: Should not be reverted.

File: libjolie/src/jolie/lang/parse/util/impl/ProgramInspectorCreatorVisitor.java [hunk 48]
```diff
@@ -367,4 +295,9 @@ public void visit( InlineTreeExpressionNode n ) {}
 	public void visit( VoidExpressionNode n ) {}
 	@Override
 	public void visit( ProvideUntilStatement n ) {}
+
+	@Override
+	public void visit(TypeChoiceDefinition n) {
```
Comment (fmontesi) [inline libjolie/src/jolie/lang/parse/util/impl/ProgramInspectorCreatorVisitor.java]: Shouldn't it do something? We are doing something (registration) for the other types.

**Comment by fmontesi** [comment]
Thanks a lot for the pull request @agwe ! I've tried to put as many comments as I could. Let me know should you have any particular questions.

**Comment by fmontesi** [comment]
Also, we should extend test/primitives/types.ol to test something about type choices, or we have no way to test that the implementation works. Do you have any example programs?

**Comment by agwe** [comment]
Hello Fabrizio,

Thank you very much for code review!
I've fixed most of the issues and sent you the new one and I'm also very
sorry for sending the request reverting your changes - it happens because I
pulled master to java8 branch without realizing that java8 is ahead.
However, I have some question regarding the previous pull request:

1. Based on your comments:
[image: Встроенное изображение 2]
[image: Встроенное изображение 1]
I haven't done anything with Type and TypeDefinition classes, considering
that it would me quite serious invasaion, changing architecture, but I can,
if you allow me to do it.

2. I'm also not sure that I have clearly understood the next comment:
[image: Встроенное изображение 3]
Now it works in the way that for types "*type choice1: int | long*" and the
type with subtype "*type choice2: void {.x: int | long}*" I got the same
error: "*Output message TypeMismatch (test_choice@ChoiceService): Invalid
native type for node #Message: expected INT, found java.lang.String*". Ant
it also works the same for native types like "*type native1: int*", "*type
native2: void {.x:int}*".
Should I fix here anything?

3. I haven't extend test suite, because I faced the next problem: running
it either from IDE, or from command line it finishes silently without any
message (or with standart "Process finished with exit code 0" in IDE). It
happens, even if I change tests making them deliberately failed - no error
message appears anyway.
Could you please explain me how to run the tests?

When I tested my solution I used the following combinations with both
positive and negative cases:
//test choice operator
type choice1: int | long

//test choice between subtypes
type choice2: void {.x: int | long}

//test choice between native and link
type str: string
type choice3: int | str


4. I have also noticed one curious thing, which is not related to pull
request and I'm not sure is it a bug or I just didn't understand how it
should work. When I tested different type combinations and passed from
client to server the following:
*  value = 1;*
*  test_choice@ChoiceService( value )( response ); *
and server had the following code:
*  main*
*  {*
* [ test_choice( request )( result )] { result = request }*
* [ test_choice_subtype ( request )( result ) ] { result.x = request.x }*
* [ test_choice_linked ( request )( result ) ] { result = request }*
*  }*
I got the message:
*INFO: [server.ol] Request-Response input operation output value
TypeMismatch (operation test_choice): Invalid native type for node
#Message: expected INT, found void*

It worked if server contained only "*test_choice( request )( result )] {
result = request }*".

Maybe I didn't understand correctly input choice semantics?

Thank you in advance.

Best Regards
Larisa

2015-08-31 14:08 GMT+03:00 Fabrizio Montesi <notifications@github.com>:

**Comment by fmontesi** [comment]
Thanks a lot for the new pull request!

About 1:
Yes, you're right. I'd like to take a shot at that first to fully realise the consequences, so for now I'll take the pull request and see if the changes required are easy. If they are, I'll just add them, otherwise we'll discuss again.

About 2:
I can't see my comment anymore and github does not show it to me, so I don't know what you're referring to unfortunately.

About 3:

The test suite is run either by using `ant test` or by running `jolie test.ol` in the `test` subdirectory. You will see that there are many tests inside that are run one after the other. You could extend the test primitives/types.ol for example.

About 4:

That sounds like a bug, but to be sure we need to see an executable program that we can test. Can you produce that and create an issue so that we can all look into it?
